docs/deployment-patterns/deployment-as-code.md

```diff
@@ -0,0 +1,111 @@
+1. **Push configuration changes to Octopus project treating the code as a source of truth**
+1. Push application packages to deployment feed
+1. Create a release in Octopus
+1. Deploy the release via Octopus to your Dev/Test environment
+1. Promote the release to other environments
+
+### Configure Octopus using code {#configure-octopus-using-code}
+
+Octopus has a comprehensive HTTP API and .NET SDK you can use to automate **everything** in Octopus - if you can do something through the user interface, you can automate it with code. You can create and update projects, variables, deployment processes, and more. Today, this is our only fully-supported **Octopus as Code** solution.
+
+There is an [open source Terraform provider for Octopus](https://github.com/MattHodge/terraform-provider-octopusdeploy), which is built on top of the Octopus HTTP API. We are using this Terraform provider ourselves for [Octopus Cloud](https://octopus.com/cloud) (our SaaS product), and we are actively contributing to it. It doesn’t cover 100% of all Octopus features yet, and the structure of the Terraform resources are subject to change. We will be building first-class support for this into the Octopus ecosystem in the future.
+
+If you want to do **Octopus configuration as code** today, we recommend using our .NET SDK which will always be supported. The Terraform provider will be a simpler, more declarative approach that we will support in the future.
+
+### Consistency and repeatability using project as code
+
+When manage your Octopus configuration as code, Octopus still makes sure your deployments are consistent and repeatable. Whenever you push a configuration change to your Octopus project via code, it's just like people using the user interface or API to make changes. When you create a release, Octopus takes a snapshot of the deployment process, variables, and packages making every deployment of that release consistent and repeatable, regardless of whether the project was configured by a person or by code.
+
+This means you can push configuration changes to Octopus as code, and get the same consistent and repeatable experience you expect for your deployments.
+
+### Introducing changes safely using project as code
+
+This is where things get a bit more difficult compared to the [scripts as code](#scripts-as-code) pattern.
+
+As a starting point, we recommend **pushing configuration changes from one specific branch into Octopus.** Using the example of git, you should only push changes from the `master` branch into Octopus, and use [Channels](/docs/deployment-process/channels/index.md) to safely introduce changes to your process and variables.
+
+The general process should look something like this, tailored to your situation:
+
+1. Create a channel to match your branch, with package version rules to enforce the integrity of the release process. _You can create channels manually, or automatically as part of your build pipeline if that suits._
```

Comment:
In Hosted v2 (https://test-hub.octopushq.com/app#/Spaces-1) we are using `branch name` as a prefix to make sure each branch gets its own set of resources in Octopus. The end goal is to use Spaces for that. Maybe we can recommend that? In this way you don't have to have one blessed branch for deployment process.

Reply:
Oh, that's an interesting idea. That's not an immediately obvious opportunity. I'll try find a way to weave the idea in.